plat-stm32mp1: bumping to pinctrl framework

etienne-lms opened the pull request:
Bump stm32_gpio driver to recently merged  pinctrl device driver probing support.

For STM32MP13 variant to comply with pinctrl support (and not panic on driver probe errors), I needed to update STM32MP13 DTS and DTSI file and 2 stm32 drivers (stm32_etzpc and stm32_rstctrl). They are part of this series.

A patch fixing a pinctrl driver probe function in under review at https://github.com/OP-TEE/optee_os/pull/6081.

etienne-lms commented:
This series it too long. I'll shorten it and rebase it.

GseoC commented on core/drivers/stm32_gpio.c:
```diff
@@ -634,11 +663,108 @@ void stm32_gpio_set_secure_cfg(unsigned int bank_id, unsigned int pin,
+			panic();
+	}
+
+	loc_conf = calloc(1, sizeof(*loc_conf) + sizeof(*pinctrl) * pin_count);
```
Is is better to do this shortcut or explicitely allocate memory for pinctrl structures in a separate calloc?

etienne-lms replied:
I think it is preferred, to lower the heap fragmentaiton.

GseoC commented on core/drivers/stm32_gpio.c:
```diff
@@ -634,11 +663,108 @@ void stm32_gpio_set_secure_cfg(unsigned int bank_id, unsigned int pin,
+	count = 0;
+	fdt_for_each_subnode(pinmux_node, fdt, pinctrl_node) {
+		int found = 0;
+
```
Can you detail the error case please ?

etienne-lms replied:
you mean why do we panic at line 800?

GseoC replied:
Yes indeed :)

etienne-lms replied:
i'll add an inline comment

etienne-lms commented on core/drivers/stm32_gpio.c:
```diff
@@ -51,6 +52,16 @@
 #define DT_GPIO_BANK_NAME0	"GPIOA"
 
+/*
+ * struct stm32_pinctrl_array - Array of pins in a pin control state
```
A "pin control state" here relates to `struct pinctrl_state`: it is a configuration set for a set of pins for a specific pin muxing use. For example a min muxing state to I2C bus wires when I2C bus is active, and other set when I2C bus is sleeping.

etienne-lms commented on core/drivers/stm32_gpio.c:
```diff
@@ -51,6 +52,16 @@
+ */
+struct stm32_pinctrl_array {
+	size_t count;
+	struct stm32_pinctrl pinctrl[];
```
no: field `pinctrl` is not a pointer to an array allocated somewhere, it is the start point of the array.

etienne-lms commented on core/drivers/stm32_gpio.c:
```diff
@@ -322,14 +338,24 @@ static int get_pinctrl_from_fdt(void *fdt, int node,
 
 			ref->bank = (uint8_t)bank;
 			ref->pin = (uint8_t)pin;
+#ifdef CFG_DRIVERS_PINCTRL
```
These switch will be shortly removed (once the [full patches series](https://github.com/OP-TEE/optee_os/pull/6050) is merged) as I expect stm32_gpio to mandate `CFG_DRIVERS_PINCTRL=y`.

GseoC replied:
Guess that's fine then

etienne-lms commented on core/drivers/stm32_gpio.c:
```diff
@@ -322,14 +338,24 @@ static int get_pinctrl_from_fdt(void *fdt, int node,
 			ref->pin = (uint8_t)pin;
+#ifdef CFG_DRIVERS_PINCTRL
+			ref->cfg.mode = mode;
+			ref->cfg.otype = opendrain ? 1 : 0;
```
i'll check that. maybe in a later P-R, to not mix a fix and this change taht prepares for pinctrl support.

etienne-lms commented on core/drivers/stm32_gpio.c:
```diff
@@ -634,11 +663,108 @@ void stm32_gpio_set_secure_cfg(unsigned int bank_id, unsigned int pin,
 }
 
+#ifdef CFG_DRIVERS_PINCTRL
+static TEE_Result stm32_pinctrl_conf_apply(struct pinconf *conf)
```
this function must comply with `struct pinctrl_ops::pinctrl_apply`

etienne-lms commented on core/drivers/stm32_gpio.c:
```diff
@@ -51,6 +52,16 @@
+/*
+ * struct stm32_pinctrl_array - Array of pins in a pin control state
+ * @count: Number of cells in @pinctrl
+ * @pinctrl: Pin controls configuration
```
I went for `Pin control configuration`

etienne-lms commented on core/drivers/stm32_gpio.c:
```diff
@@ -322,14 +338,24 @@ static int get_pinctrl_from_fdt(void *fdt, int node,
+			ref->cfg.af = alternate;
+#else
 			ref->active_cfg.mode = mode;
 			ref->active_cfg.otype = opendrain ? 1 : 0;
```
I on't change this as this code will soon be removed.

etienne-lms commented:
for info, CI checkpatch error reports are false positives (build error trace messages dumped in commit messages).

etienne-lms commented:
review comments addressed

etienne-lms commented:
may I squash the fixup commits?

GseoC commented on core/arch/arm/plat-stm32mp1/shared_resources.c:
```diff
@@ -371,54 +371,66 @@ void stm32mp_register_non_secure_gpio(unsigned int bank, unsigned int pin)
-		return;
+	res = stm32_gpio_pinctrl_bank_pin(pinctrl, NULL, NULL, &count);
+	if (res || !count)
+		goto out;
```
might as well return TEE_SUCCESS here as it precedes the allocation.

GseoC commented on core/arch/arm/plat-stm32mp1/shared_resources.c:
```diff
@@ -371,54 +371,66 @@ void stm32mp_register_non_secure_gpio(unsigned int bank, unsigned int pin)
-		return;
+	res = stm32_gpio_pinctrl_bank_pin(pinctrl, NULL, NULL, &count);
+	if (res || !count)
+		goto out;
```
ditto

GseoC commented:
> may I squash the fixup commits?

For clearer view, yes please

etienne-lms commented:
Fixup commits squashed + addressed comment on `stm32mp_register_secure_pinctrl()`/`stm32mp_register_non_secure_pinctrl()`.

GseoC commented:
Thank you,

`Acked-by: Gatien Chevallier <gatien.chevallier@foss.st.com>`

etienne-lms commented:
tag applied
thanks

etienne-lms commented:
I need to rebase due to https://github.com/OP-TEE/optee_os/pull/6098 being recently merged.

etienne-lms commented:
@GseoC, I added a fixup commit to "drivers: stm32_gpio: support CFG_DRIVERS_PINCTRL", to address the rebase issue.
Could you please confirm (or infirm) your A-b tag with this change?

jforissier commented:
I have restarted IBART which failed on a socket test (rcu_sched stalls), looks unrelated to this PR.

etienne-lms commented:
thanks

GseoC commented:
Seems fine, tag applies

etienne-lms commented:
Some changes in the series, to finalize bumping _stm32_gpio.c_ to GPIO and PINCTRL frameworks:
- 1st commit is unmodified regarding previous review. @GseoC review tag is already applied to the commit.
"drivers: stm32_gpio: support CFG_DRIVERS_PINCTRL"

- The 5 next commit were review but I changed added functions prototypes to return `void` instead of a useless `TEE_Result`. @GseoC, I removed your review tag from those commits:
"drivers: stm32_gpio: add helper function stm32_gpio_pinctrl_bank_pin()" 
"drivers: stm32_gpio: add helper function stm32_pinctrl_set_secure_cfg()"
"plat-stm32mp1: shared_resources: support CFG_DRIVERS_PINCTRL"
"drivers: stm32_uart: support CFG_DRIVERS_PINCTRL" 
"drivers: stm32_i2c: support CFG_DRIVERS_PINCTRL"

- I added 5 new commits in the series, to finalize the evolution: enforce `CFG_DRIVERS_PINCTRL=y` and remove deprecated functions:
"plat-stm32mp1: conf: enable CFG_DRIVERS_PINCTRL"
"drivers: stm32_gpio: remove cases when CFG_DRIVERS_PINCTRL is disabled"
"drivers: stm32_uart: remove cases when CFG_DRIVERS_PINCTRL is disabled" 
"drivers: stm32_i2c: remove cases when CFG_DRIVERS_PINCTRL is disabled"
"drivers: stm32_gpio: move definitions to source file"

etienne-lms commented:
Changed title has the seire finalizes stm32mp1 upgrades to pinctrl framework.
Checkpatch failure on a typo in a commit. I'll the 2 fixed after the review.

etienne-lms commented on core/drivers/stm32_gpio.c:
```diff
@@ -56,6 +57,61 @@
 
 #define DT_GPIO_BANK_NAME0	"GPIOA"
 
+#define GPIO_MODE_INPUT		0x0
```
Ok., i'll fix but I wonder if it's really needed?

etienne-lms commented on core/drivers/stm32_gpio.c:
```diff
@@ -56,6 +57,61 @@
+ * @bank: GPIO bank identifier as assigned by the platform
+ * @pin: Pin number in the GPIO bank
+ * @cfg: Pin configuration
+ * @active_cfg: Configuration in active state
```
right, thanks

etienne-lms commented on core/drivers/stm32_gpio.c:
```diff
@@ -902,6 +902,29 @@ void stm32_gpio_pinctrl_bank_pin(struct pinctrl_state *pinctrl,
+	size_t conf_index = 0;
+
+	if (!pinctrl)
+		return;
```
if there is no pinctrl, there is no pin to configure so returning straight is fine, IMO.

GseoC commented:
I do not catch why the code style check fails, I guess it did not restart

`Acked-by: Gatien Chevallier <gatien.chevallier@foss.st.com>`

jforissier commented:
> I do not catch why the code style check fails, I guess it did not restart

It did, but the way the code style check works might be a bit counter-intuitive :wink: It is actually run twice:
- once for each commit in the PR, and
- once on the global diff (all commits squashed together if you wish)
The rationale is, we want to check the code but also the commit descriptions, so each commit is checked. When fixes are done as fixup commits (as opposed to force-pushing with corrected commits), the initial failure remain visible. However thanks to the second phase (called `checkdiff` in the log), you can be confident that the code style issues will be gone when squashing the fixups.

GseoC commented:
> > I do not catch why the code style check fails, I guess it did not restart
> 
> It did, but the way the code style check works might be a bit counter-intuitive wink It is actually run twice:
> 
> * once for each commit in the PR, and
> * once on the global diff (all commits squashed together if you wish)
> 
> The rationale is, we want to check the code but also the commit descriptions, so each commit is checked (the `checkpatch` lines in the log). When fixes are done as fixup commits (as opposed to force-pushing with corrected commits), the initial failure remain visible. However thanks to the second phase (called `checkdiff` in the log), you can be confident that the code style issues will be gone when squashing the fixups.

Thanks for the explanation :)

etienne-lms commented:
I'll sqaush the fixup commits. Once done, checkpatch should be happy.

etienne-lms commented:
Thakns for the review.
Fixup commits squashed.
Review tag applied.

I also removed a suprious inline comment in comment "drivers: stm32_i2c: remove cases when CFG_DRIVERS_PINCTRL is disabled" (below). @GseoC, I preserved your review tag on that commit, I think it still applies :)
```patch
@@ -113,7 +113,6 @@ struct i2c_cfg {
  * @sec_cfg: I2C registers configuration storage
  * @pinctrl: Pin control configuration for the I2C bus in active state
  * @pinctrl_sleep: Pin control configuration for the I2C bus in standby state
- * @pinctrl_count: Number of PINCTRLs elements
  */
 struct i2c_handle_s {
        struct io_pa_va base;
